File: editor/src/com/mbrlabs/mundus/history/CommandHistory.java
/*
 * Copyright (c) 2016. See AUTHORS file.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.mbrlabs.mundus.history;

import com.badlogic.gdx.utils.Array;
import com.mbrlabs.mundus.utils.Log;

/**
 * @author Marcus Brummer
 * @version 07-02-2016
 */
public class CommandHistory {

    public static final int DEFAULT_LIMIT = 20;

    private int limit;
    private int pointer;
    private Array<Command> commands;

    public CommandHistory(int limit) {
        this.limit = limit;
        commands = new Array<>(limit);
        pointer = -1;
    }

    public int add(Command command) {
        if(pointer + 1 < size()) {
            commands.removeRange(pointer + 1, commands.size - 1);
            commands.add(command);
            pointer++;
            Log.debug("HISTORY ADD", "pointer < size() ==> pointer: " + pointer);
        } else {
            if(size() == limit) {
                commands.removeIndex(0);
                commands.add(command);
                Log.debug("HISTORY ADD", "pointer == limit ==> pointer: " + pointer);
            } else {
                commands.add(command);
                pointer++;
                Log.debug("HISTORY ADD", "pointer < limit pointer == size ==> pointer: " + pointer);
            }
        }

        return pointer;
    }

    public int goBack() {
        if(pointer >= 0) {
            commands.get(pointer).undo();
            pointer--;
        }
        Log.debug("HISTORY BACK", "pointer: " + pointer);

        return pointer;
    }

    public int goForward() {
        if(pointer < commands.size - 1 ) {
            pointer++;
            commands.get(pointer).execute();
        }
        Log.debug("HISTORY FORWARD", "pointer: " + pointer);

        return pointer;
    }

    public int getPointer() {
        return pointer;
    }

    public void clear() {
        commands.clear();
        pointer = 0;
    }

    public int size() {
        return commands.size;
    }

}

File: editor/src/main/java/com/mbrlabs/mundus/editor/history/CommandHistory.kt
/*
 * Copyright (c) 2016. See AUTHORS file.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.mbrlabs.mundus.editor.history

import com.badlogic.gdx.utils.Array

/**
 * Add commands to undo/redo a previously called process.
 *
 * @author Marcus Brummer
 * @version 07-02-2016
 */
class CommandHistory(private val limit: Int) {

    private var pointer: Int = 0
    private val commands: Array<Command>

    init {
        commands = Array<Command>(limit)
        pointer = -1
    }

    fun add(command: Command): Int {
        if (size() == 0) {
            commands.add(command)
            pointer++
            return pointer
        }

        if (pointer < size() - 1) {
            removeCommands(pointer + 1, commands.size - 1)
            commands.add(command)
            pointer++
        } else {
            if (size() == limit) {
                removeCommand(0)
                commands.add(command)
            } else {
                commands.add(command)
                pointer++
            }
        }

        return pointer
    }

    private fun removeCommand(index: Int) {
        commands.get(index).dispose()
        commands.removeIndex(index)
    }

    private fun removeCommands(from: Int, to: Int) {
        for (i in from..to) {
            commands.get(i).dispose()
        }

        commands.removeRange(from, to)
    }

    fun goBack(): Int {
        if (pointer >= 0) {
            commands.get(pointer).undo()
            pointer--
        }

        return pointer
    }

    fun goForward(): Int {
        if (pointer < commands.size - 1) {
            pointer++
            commands.get(pointer).execute()
        }

        return pointer
    }

    fun clear() {
        for (c in commands) {
            c.dispose()
        }
        commands.clear()
        pointer = -1
    }

    fun size(): Int {
        return commands.size
    }

    companion object {

        val DEFAULT_LIMIT = 50
    }

}
